### PR TITLE
16879 detectoverflow configurable collision boundary

### LINKED ISSUE #16879: Floating UI: Add `detectOverflow` configurable collision boundary
In certain situations consumers may need to customize the boundary/element used to detect collisions. One example of this might be the modal - components like Popover may need to reflow based on the bounds of the modal content container, and not the viewport. 

To this we can implement the [detectOverflow](https://floating-ui.com/docs/detectoverflow) as a middleware 

Eventually we could also
* add props for consumers to pass in their own padding
* add this prop to other components using floating-ui autoAlign
* ensure this is configurable from anywhere in their react tree #17003

### PR DESCRIPTION
**Pull request by preetibansalui**
Part of #16879 

Added `hide` from Floating-ui to Dropdown component so it will hide the floating dropdown if the reference (Dropdown Trigger point) is not visible.

#### Testing / Reviewing

1. Check the ExperimentalAutoAlignBoundaryTest Story in Dropdown. 
2. Open the dropdown and scroll it, the moment 'Option 1' hide from the screen, it should hide the floating dropdown options as well.

### CODE REVIEW

File: packages/react/src/components/Dropdown/Dropdown.tsx [hunk 1]
```diff
@@ -299,13 +300,31 @@ const Dropdown = React.forwardRef(
...
+                ? 'hidden'
+                : 'visible',
+            });
+        }
+      });
+    };
+
     useEffect(() => {
       if (autoAlign) {
         Object.keys(floatingStyles).forEach((style) => {
```
Comment (tay1orjones) [inline packages/react/src/components/Dropdown/Dropdown.tsx]: If we get the middleware working with the hook, the visibility style could be applied by adding it to `floatingStyles` here before we iterate over it.

```suggestion
      if (autoAlign) {
        Object.keys({...floatingStyles, visibility: middlewareData.hide?.referenceHidden ? 'hidden' : 'visible'}).forEach((style) => {
```

Or maybe on the preceding line so it's less dense

**Review by tay1orjones — COMMENTED** [review]
Did we want to try and see if we can use the `hide()` middleware directly with the useFloating hook? https://floating-ui.com/docs/hide

**Comment by preetibansalui** [comment]
> Did we want to try and see if we can use the `hide()` middleware directly with the useFloating hook? https://floating-ui.com/docs/hide

Hey @tay1orjones, Thanks for the suggestion. It works fine in useFloating hook as well so kept it in hook to avoid extra code.

**Review by guidari — APPROVED** [review]
Awesome! LGTM! 🚀

**Review by tay1orjones — APPROVED** [review]
LGTM once the test story is removed

**Pull request merged**